### PR TITLE
add support outputting json format results to a file. (issue: 101)

### PR DESCRIPTION
**Pull request by iyoshiha**
## Details of the Changes

- Added an output flag option for json format. 
- Added tests
- Refactored standard output format to add outputting results to a file feature easily in the feature (if it's required.) 
  
## Issue ticket number and link
Closes https://github.com/Boeing/config-file-validator/issues/101

I'll add this feature for JUnit if this PR is merged into main.

### CODE REVIEW

File: cmd/validator/validator.go [hunk 1]
```diff
@@ -66,10 +69,11 @@ func validatorUsage() {
 // will return with exit = 1
 func getFlags() (validatorConfig, error) {
 	flag.Usage = validatorUsage
+	depthPtr := flag.Int("depth", 0, "Depth of recursion for the provided search paths. Set depth to 0 to disable recursive path traversal")
 	excludeDirsPtr := flag.String("exclude-dirs", "", "Subdirectories to exclude when searching for configuration files")
-	reportTypePtr := flag.String("reporter", "standard", "Format of the printed report. Options are standard, json and junit")
 	excludeFileTypesPtr := flag.String("exclude-file-types", "", "A comma separated list of file types to ignore")
-	depthPtr := flag.Int("depth", 0, "Depth of recursion for the provided search paths. Set depth to 0 to disable recursive path traversal")
+	outputPtr := flag.String("output", "", "Destination to a file to output results")
+	reportTypePtr := flag.String("reporter", "standard", "Format of the printed report. Options are standard and json")
```
Comment (iyoshiha) [inline cmd/validator/validator.go:76]: refactor: alphabetized

File: test/output/example/good.json
Comment (iyoshiha) [inline test/output/example/good.json:1]: this file is used for `example/result.json`

File: pkg/reporter/stdout_reporter.go
Comment (iyoshiha) [inline pkg/reporter/stdout_reporter.go:1]: refactor : easily available to add an output results to a file option.

File: validator
Comment (kehoecj) [inline validator:1]: Looks like this accidentally got pushed

Comment (iyoshiha) [reply validator:1]: I removed the binary file. Thank you for pointing out the mistake!

File: cmd/validator/validator.go [hunk 2]
```diff
@@ -17,6 +17,8 @@ optional flags:
     	Subdirectories to exclude when searching for configuration files
   -exclude-file-types string
     	A comma separated list of file types to ignore
+  -output
+  		Destination of a file to outputting results
```
Comment (jd4235) [inline cmd/validator/validator.go:21]: Indent is off here. Ensure it aligns with the rest of the descriptions please.

File: pkg/reporter/reporter_test.go [hunk 3]
```diff
@@ -130,3 +136,128 @@ func Test_junitReport(t *testing.T) {
...
+	type want struct {
+		data []byte
+		err  assert.ErrorAssertionFunc
+	}
+
+	tests := map[string]struct {
+		args args
+		want want
+	}{
+		"Normal/Output results to a file named 'result.json' (default name)": {
```
Comment (jd4235) [inline pkg/reporter/reporter_test.go:167]: Please add a test where the string is empty to ensure your response is appropriate.

File: cmd/validator/validator.go [hunk 4]
```diff
@@ -148,7 +153,7 @@ func mainInit() int {
 	// since the exclude dirs are a comma separated string
 	// it needs to be split into a slice of strings
 	excludeDirs := strings.Split(*validatorConfig.excludeDirs, ",")
-	reporter := getReporter(validatorConfig.reportType)
+	reporter := getReporter(validatorConfig.reportType, validatorConfig.output)
```
Comment (jd4235) [inline cmd/validator/validator.go:156]: What is supposed to happen if the output option if specified and the string is empty?  Be sure to add a test for that and set the appropriate response.

File: pkg/reporter/json_reporter.go [hunk 5]
```diff
@@ -64,6 +73,12 @@ func (jr JsonReporter) Print(reports []Report) error {
 		return err
 	}
 
-	fmt.Println(string(jsonBytes))
+	jsonBytes = append(jsonBytes, '\n')
+	fmt.Print(string(jsonBytes))
+
+	if jr.outputDest != "" {
```
Comment (jd4235) [inline pkg/reporter/json_reporter.go:79]: Where does outputDest get initialized to an empty string?

File: pkg/reporter/writer.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package reporter
+
+import (
+	"fmt"
+	"os"
+)
+
+func outputResultsToFile(outputDest, extension string, results []byte) error {
```
Comment (jd4235) [inline pkg/reporter/writer.go:8]: Need comment header. Also I didn't see a unit test for this.  Please add.

File: pkg/reporter/json_reporter.go (hunk 5 above)
Comment (iyoshiha) [reply pkg/reporter/json_reporter.go:79]: [This line](https://github.com/Boeing/config-file-validator/blob/d49614c2fd63231b4e3e07eca4d9d6882fba1c4c/cmd/validator/validator.go#L75) initializes outputPtr variable as an empty string  when `-output` flag is not provided. When `reporter` is initialized, outputPtr value is assigned to the reporter filed `outputDest`.

File: cmd/validator/validator.go (hunk 4 above)
Comment (iyoshiha) [reply cmd/validator/validator.go:156]: I added some test cases. Please check them. ae9c479a9e20d6e99ce73ecb9b3a2c61863468d6

File: pkg/reporter/reporter_test.go (hunk 3 above)
Comment (iyoshiha) [reply pkg/reporter/reporter_test.go]: Thank you for pointing out that. I added an empty cases. 20f3c63d82443f8b71a8c8929bb15d50c4563a2f

File: cmd/validator/validator.go (hunk 2 above)
Comment (iyoshiha) [reply cmd/validator/validator.go:21]: fixed it. Thanks.

**Comment by iyoshiha** [comment]
@jd4235 
i fixed and add some tests. Please re-review again.

**Comment by jd4235** [comment]
> @jd4235 i fixed and add some tests. Please re-review again.

LGTM.

File: cmd/validator/validator.go (hunk 4 above)
Comment (jd4235) [reply cmd/validator/validator.go:156]: Looks good. Thanks.

**Comment by kehoecj** [comment]
@iyoshiha JUnit changes were merged so you'll need to add output functionality for junit. Currently when I run the command `.\validator.exe --reporter junit --output results.xml .\test\` the `results.xml` file is not created

**Comment by iyoshiha** [comment]
@kehoecj 
May I ask you about test coverage please? 
i added output function for junit but [go update coverage badge](https://github.com/Boeing/config-file-validator/actions/runs/7257248336/job/19770776221?pr=104) failed. this says 
```
Coverage threshold: 95 %
Current test coverage : 94.9 %
```
test coverage is 0.1% below. What can I do to increase test coverage?

**Comment by iyoshiha** [comment]
@kehoecj 
i added output function for junit. please review again.

**Comment by jd4235** [comment]
Thanks for adding those extra tests. Looks ready to me.

**Comment by kehoecj** [comment]
Merged! Thank you @iyoshiha for the contribution!
